superduperdb/components/graph.py

```diff
@@ -0,0 +1,332 @@
+import dataclasses as dc
+import typing as t
+
+import networkx as nx
+
+from superduperdb import Schema
+from superduperdb.backends.query_dataset import QueryDataset
+from superduperdb.components.model import Signature, _Predictor
+
+
+class GraphModel:
+    def __init__(self, name, signature):
+        self.identifier = name
+        self.inputs = []
+        self.signature = signature
+
+    def predict_one(self, x):
+        return x
+
+    def predict(self, dataset):
+        outs = []
+        for i in dataset:
+            args = i
+            outs.append(self.predict_one(*args))
+        return outs
+
+
+@dc.dataclass(kw_only=True)
+class Graph(_Predictor):
+    '''
+    Represents a directed acyclic graph composed of interconnected model nodes.
+
+    This class enables the creation of complex predictive models
+    by defining a computational graph structure where each node
+    represents a predictive model. Models can be connected via edges
+    to define dependencies and flow of data.
+
+    The predict() method executes predictions through the graph, ensuring
+    correct data flow and handling of dependencies
+
+    Example:
+    >>  g = Graph(
+    >>    identifier='simple-graph', input=model1, outputs=[model2], signature='*args'
+    >>  )
+    >>  g.connect(model1, model2)
+    >>  assert g.predict_one(1) == [(4, 2)]
+
+    '''
+
+    models: t.List[_Predictor] = dc.field(default_factory=list)
+    edges: t.List[t.Tuple[str, str, t.Tuple[int, str]]] = dc.field(default_factory=list)
+    input: _Predictor
+    outputs: t.List[t.Union[str, _Predictor]] = dc.field(default_factory=list)
+    _DEFAULT_ARG_WEIGHT: t.ClassVar[tuple] = (None, Signature.singleton)
+    signature: str = Signature.args_kwargs  # type: ignore[misc]
+    type_id: t.ClassVar[str] = 'graph'
+
+    def __post_init__(self, artifacts):
+        self.G = nx.DiGraph()
+        self.nodes = {}
+        self._node_output_cache = {}
+        self.version = 0
+        self._db = None
+
+        assert all([isinstance(o, _Predictor) for o in self.outputs])
+
+        self.output_schema = Schema(
+            identifier=self.identifier,
+            fields={k.identifier: k.datatype for k in self.outputs},
+        )
+
+        self.outputs = [
+            o.identifier if isinstance(o, _Predictor) else o for o in self.outputs
+        ]
+        self.input.signature = Signature.args
+
+        self.connect_root(self.input)
+
+        # Load the models and edges into a di graph
+        models = {m.identifier: m for m in self.models}
+        if self.edges and models:
+            for connection in self.edges:
+                u, v, on = connection
+                self.connect(
+                    models[u],
+                    models[v],
+                    on=on,
+                    update_edge=False,
+                )
+        super().__post_init__(artifacts=artifacts)
+
+    def connect_root(self, v: _Predictor):
```

Comment:
Not necessary. Root can just be input node.

Reply:
leftover from previous algorithm 
removed now